### PR TITLE
added typescript definitions

### PR DESCRIPTION
**Pull request by srghma**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+   * Merging with a customizable merge strategy
+   */
+  const strategy: (options: StrategyMap) => ConfigurationJoiner
+
+  /**
+   * Merging with a customizable smart merge strategy
+   */
+  const smartStrategy: (options: StrategyMap) => ConfigurationJoiner
+
+  interface NonstandartConfigurationJoiner {
```
Comment (bebraw) [inline index.d.ts]: Typo. standart -> standard.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    (...configuration: any[]): Configuration
+    (configuration: any[]): Configuration
+  }
+
+  /**
+   * Merging with a smart merge strategy
+   */
+  const smart: NonstandartConfigurationJoiner
```
Comment (bebraw) [inline index.d.ts]: Same here.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  /**
+   * Merging with a multiple merge strategy
+   */
+  const multiple: NonstandartConfigurationJoiner
```
Comment (bebraw) [inline index.d.ts]: And here.

File: package.json [hunk 4]
```diff
@@ -12,13 +12,15 @@
     "preversion": "npm run test:lint && npm run build && npm test && git commit --allow-empty -am \"Update lib\""
   },
   "main": "lib/index.js",
+  "typings": "index.d.ts",
```
Comment (bebraw) [inline package.json]: Would it make sense to port the whole code to TypeScript and drop the type definition file?

Comment (bebraw) [reply package.json]: We could still generate JS + typedefs based on that.

Comment (BjornMelgaard) [reply package.json]: I'm sorry, i'm not using this library often, I'm just wanted to learn ts

Comment (bebraw) [reply package.json]: No worries. 👍

**Comment by bebraw** [comment]
@rockmacaca Can you comment on this?

**Comment by rockmacaca** [comment]
I'd side with just porting the thing over to TS, especially since you're already set up with running everything through Babel. I've found that most non TS libraries I've used have a lot of trouble maintaining their own type definitions. I'm not familiar with how publishing type definitions works, but I think it's a pretty simple process.

**Comment by bebraw** [comment]
Yeah, let's wait for a full TS port. Agreed.

I might do it myself in two weeks unless someone else does it first. This PR is a great starting point.

**Comment by leebenson** [comment]
@bebraw - just wondering if the TS port is still happening?

**Comment by bebraw** [comment]
@leebenson I would love to do a full port but I'm not well-funded atm and time is quite short with other things. That said, if someone wants to contribute the work, I don't mind merging. Types can be made more strict as needed. Now the definition is quite loose sometimes.

**Comment by leebenson** [comment]
understood @bebraw, I appreciate all your hard work to date. I'm pretty swamped myself atm, but if I get any time, I might take a stab at a TS port.

**Pull request closed without merging**